presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java

```diff
@@ -123,80 +125,86 @@ private PlanNodeStatsEstimate filterForFalseExpression()
                                             .setDistinctValuesCount(0.0)
                                             .setNullsFraction(NaN).build()));
 
-            return falseStatsBuilder.setOutputRowCount(0.0).build();
+            return Optional.of(falseStatsBuilder.setOutputRowCount(0.0).build());
         }
 
         @Override
-        protected PlanNodeStatsEstimate visitNotExpression(NotExpression node, Void context)
+        protected Optional<PlanNodeStatsEstimate> visitNotExpression(NotExpression node, Void context)
         {
-            return differenceInStats(input, process(node.getValue()));
+            return process(node.getValue()).map(childStats -> differenceInStats(input, childStats));
         }
 
         @Override
-        protected PlanNodeStatsEstimate visitLogicalBinaryExpression(LogicalBinaryExpression node, Void context)
+        protected Optional<PlanNodeStatsEstimate> visitLogicalBinaryExpression(LogicalBinaryExpression node, Void context)
         {
-            PlanNodeStatsEstimate leftStats = process(node.getLeft());
-            PlanNodeStatsEstimate andStats = new FilterExpressionStatsCalculatingVisitor(leftStats, session, types).process(node.getRight());
-
+            Optional<PlanNodeStatsEstimate> leftStats = process(node.getLeft());
+            Optional<PlanNodeStatsEstimate> andStats = new FilterExpressionStatsCalculatingVisitor(leftStats.orElse(filterForUnknownExpression()), session, types).process(node.getRight());
```

Comment:
this section is kind of confusing because you need to first realize that you set the value for left stats to always have a value, so if `andStats` are empty  that means the right stats are unknown.   I'm not sure I have a better suggestion though.  Maybe explicitly first computing left and  right stats would make it clearer?